skyportal/handlers/api/plan.py

```diff
@@ -0,0 +1,450 @@
+import datetime
+import os
+import copy
+
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.orm import joinedload
+from astropy import time
+import ephem
+import gwemopt.utils
+import gwemopt.skyportal
+import gwemopt.gracedb
+import gwemopt.rankedTilesGenerator
+import gwemopt.waw
+import gwemopt.lightcurve
+import gwemopt.coverage
+import gwemopt.efficiency
+import gwemopt.plotting
+import gwemopt.tiles
+import gwemopt.segments
+import gwemopt.catalog
+import numpy as np
+
+from baselayer.app.access import auth_or_token
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    Instrument,
+    Localization,
+    ObservingPlan,
+    PlannedObservation,
+    Telescope,
+)
+
+plan_args_default = {
+    'filt': ['g', 'r', 'g'],
+    'exposuretimes': [300.0, 300.0, 300.0],
+    'doReferences': False,
+    'doUsePrimary': True,
+    'doBalanceExposure': False,
+    'doDither': False,
+    'usePrevious': False,
+    'doCompletedObservations': False,
+    'doPlannedObservations': False,
+    'cobs': [None, None],
+    'schedule_type': 'greedy',
+    'filterScheduleType': 'block',
+    'airmass': 2.5,
+    'schedule_strategy': 'tiling',
+    'mindiff': 30.0 * 60.0,
+    'doMaxTiles': False,
+    'max_nb_tiles': 1000,
+    'doRASlice': True,
+    'raslice': [0.0, 24.0],
+}
+
+
+class PlanHandler(BaseHandler):
+    @auth_or_token
+    def post(self):
+
+        data = self.get_json()
+
+        telescope_name = data.get('telescope_name')
+        instrument_name = data.get('instrument_name')
+        dateobs = data.get('dateobs')
+        localization_name = data.get('localization_name')
+
+        telescope = (
+            Telescope.query_records_accessible_by(
+                self.current_user,
+            )
+            .filter(
+                Telescope.name == telescope_name,
+            )
+            .one()
```

Comment:
probably better to use `first()` so we can gracefully catch when there's no match and manually return an error message, rather than having an uncaught exception add a messy TB to the logs

Reply:
yes!